core/src/toga/dialogs.py

```diff
@@ -0,0 +1,236 @@
+from __future__ import annotations
+
+import asyncio
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+from toga.platform import get_platform_factory
+
+if TYPE_CHECKING:
+    from toga.window import Window
+
+
+__all__ = [
+    "InfoDialog",
+    "QuestionDialog",
+    "ConfirmDialog",
+    "ErrorDialog",
+    "StackTraceDialog",
+    "SaveFileDialog",
+    "OpenFileDialog",
+    "SelectFolderDialog",
+]
+
+
+class Dialog:
+    """A base class for dialogs.
+
+    These classes are not displayed directly. To use them, pass a
+    :class:`~toga.dialogs.Dialog` instance to :meth:`~toga.Window.dialog()` (for a
+    window-modal dialog), or :meth:`~toga.App.dialog()` for an app-level dialog.
+    """
+
+    def _show(self, window: Window | None) -> asyncio.Future:
+        """Display the dialog and return the user's response.
+
+        :param window: The window for which the dialog should be modal; or ``None`` for
+            an app-level dialog.
+        :returns: A future capturing the user's response to the dialog
+        """
+        future = asyncio.Future()
+        self._impl.show(window, future)
+        return future
+
+
+class InfoDialog(Dialog):
+    def __init__(self, title: str, message: str):
+        """Ask the user to acknowledge some information.
+
+        Presents as a dialog with a single "OK" button to close the dialog.
+
+        Returns a response of ``None``.
+
+        :param title: The title of the dialog window.
+        :param message: The message to display.
+        """
+        self.factory = get_platform_factory()
+        self._impl = self.factory.dialogs.InfoDialog(title=title, message=message)
+
+
+class QuestionDialog(Dialog):
+    def __init__(self, title: str, message: str):
+        """Ask the user a yes/no question.
+
+        Presents as a dialog with "Yes" and "No" buttons.
+
+        Returns a response of ``True`` when the "Yes" button is pressed, ``False`` when
+        the "No" button is pressed.
+
+        :param title: The title of the dialog window.
+        :param message: The question to be answered.
+        """
+        self.factory = get_platform_factory()
+        self._impl = self.factory.dialogs.QuestionDialog(title=title, message=message)
+
+
+class ConfirmDialog(Dialog):
+    def __init__(self, title: str, message: str):
+        """Ask the user to confirm if they wish to proceed with an action.
+
+        Presents as a dialog with "Cancel" and "OK" buttons (or whatever labels are
+        appropriate on the current platform), bound to a window.
```

Comment:
Since the whole point of the PR was to enable dialogs that are not bound to a window, I'm not sure why this text was added.

```suggestion
        appropriate on the current platform).
```

Reply:
And likewise for most of the other dialogs.

Reply:
🤦 Oversight on my part. I updated the docs in the first draft when I was preserving the window-based function API.